// @sap/approuter/lib/middleware/whitelist-service-middleware.js
'use strict';

var URL = require('url');
var querystring = require('querystring');
var whitilistUtils = require('../utils/whitelist-utils');


module.exports = function (req, res, next) {
  if (req.method !== 'GET') {
    return next();
  }

  var whitelist = req.routerConfig.clickjackCheckWhitelist;
  if (!whitelist) {
    return sendServiceDisabledResult(res);
  }

  var parentOrigin = getParentOrigin(req);
  if (typeof parentOrigin !== 'string') {
    return sendResult(res, 'null', false);
  }

  var isMatching = whitilistUtils.matchWhitelist(whitelist, parentOrigin);
  sendResult(res, parentOrigin, isMatching);
};

function getParentOrigin(req) {
  var requestUrl = URL.parse(req.url);
  if (requestUrl.query) {
    return querystring.parse(requestUrl.query)['parentOrigin'];
  }
  return null;
}

function sendServiceDisabledResult(res) {
  send(res, {
    version: '1.0',
    active: false,
    framing: true
  });
}

function sendResult(res, origin, isMatching) {
  send(res, {
    version: '1.0',
    active: true,
    origin: origin,
    framing: isMatching
  });
}

function send(res, result) {
  res.writeHead(200, { 'content-type': 'application/json' });
  return res.end(JSON.stringify(result));
}
