// @sap/xssec/lib/constants.js
// Scope Prefix:
Object.defineProperty(exports, "XSAPPNAMEPREFIX", {
    value:        '$XSAPPNAME.',
    enumerable:   true,
    writable:     false,
    configurable: false
});

// Namespaces:
Object.defineProperty(exports, "SYSTEM", {
    value:        'System',
    enumerable:   true,
    writable:     false,
    configurable: false
});

// Names:
Object.defineProperty(exports, "JOBSCHEDULER", {
    value:        'JobScheduler',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "HDB", {
    value:        'HDB',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "GRANTTYPE_AUTHCODE", {
    value:        'authorization_code',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "GRANTTYPE_PASSWORD", {
    value:        'password',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "GRANTTYPE_CLIENTCREDENTIAL", {
    value:        'client_credentials',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "GRANTTYPE_USERTOKEN", {
    value:        'user_token',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "GRANTTYPE_SAML2BEARER", {
    value:        'urn:ietf:params:oauth:grant-type:saml2-bearer',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "TYPE_USER_TOKEN", {
    value:        0,
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "TYPE_CLIENT_CREDENTIALS_TOKEN", {
    value:        1,
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "KEYCACHE_DEFAULT_TOKENKEY_PATH", {
    value:        'token_keys',
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "KEYCACHE_DEFAULT_CACHE_ENTRY_EXPIRATION_TIME_IN_MINUTES", {
    value:        15,
    enumerable:   true,
    writable:     false,
    configurable: false
});

Object.defineProperty(exports, "KEYCACHE_DEFAULT_CACHE_SIZE", {
    value:        100,
    enumerable:   true,
    writable:     false,
    configurable: false
});
