// @sap/e2e-trace/lib/dsr/Record.js
'use strict';

var Passport = require('../passport/Passport');

module.exports = Record;

function Record(req) {
  this.action = buildUrl(req);

  this.receivedBytes = -1;
  this.sentBytes = -1;
  this.respTime = -1;

  this.transId = extractTransactionId(req);
  this.userId = (req.user && req.user.id) || 'n.a.';
}

function buildUrl(req) {
  var headers = req.headers;
  var protocol = headers['x-forwarded-proto'] || (req.connection.encrypted ? 'https' : 'http');
  var host = headers['x-forwarded-host'] || headers.host;
  var path = req.originalUrl || req.url;
  return protocol + '://' + host + path;
}

function extractTransactionId(req) {
  var incomigPassport = req.headers[Passport.HEADER_NAME];
  var passport = new Passport(incomigPassport);
  return passport.readUniqueIdentifiers().transactionID;
}

Record.prototype.setStartTime = function (startTime) {
  this.startTime = startTime;
};

Record.prototype.setHttpStats = function (stats) {
  this.receivedBytes = stats.req.bytes;
  this.sentBytes = stats.res.bytes;
  this.respTime = stats.time;
};

Record.prototype.log = function () {
  console.log(JSON.stringify(this)); // eslint-disable-line no-console
};
