// ajv/lib/dotjs/index.js
'use strict';

//all requires must be explicit because browserify won't work with dynamic requires
module.exports = {
  '$ref': require('ajv/lib/dotjs/ref'),
  allOf: require('ajv/lib/dotjs/allOf'),
  anyOf: require('ajv/lib/dotjs/anyOf'),
  '$comment': require('ajv/lib/dotjs/comment'),
  const: require('ajv/lib/dotjs/const'),
  contains: require('ajv/lib/dotjs/contains'),
  dependencies: require('ajv/lib/dotjs/dependencies'),
  'enum': require('ajv/lib/dotjs/enum'),
  format: require('ajv/lib/dotjs/format'),
  'if': require('ajv/lib/dotjs/if'),
  items: require('ajv/lib/dotjs/items'),
  maximum: require('ajv/lib/dotjs/_limit'),
  minimum: require('ajv/lib/dotjs/_limit'),
  maxItems: require('ajv/lib/dotjs/_limitItems'),
  minItems: require('ajv/lib/dotjs/_limitItems'),
  maxLength: require('ajv/lib/dotjs/_limitLength'),
  minLength: require('ajv/lib/dotjs/_limitLength'),
  maxProperties: require('ajv/lib/dotjs/_limitProperties'),
  minProperties: require('ajv/lib/dotjs/_limitProperties'),
  multipleOf: require('ajv/lib/dotjs/multipleOf'),
  not: require('ajv/lib/dotjs/not'),
  oneOf: require('ajv/lib/dotjs/oneOf'),
  pattern: require('ajv/lib/dotjs/pattern'),
  properties: require('ajv/lib/dotjs/properties'),
  propertyNames: require('ajv/lib/dotjs/propertyNames'),
  required: require('ajv/lib/dotjs/required'),
  uniqueItems: require('ajv/lib/dotjs/uniqueItems'),
  validate: require('ajv/lib/dotjs/validate')
};
