// @sap/e2e-trace/lib/passport/field-descriptors.js
'use strict';

var bufferUtil = require('./buffer-utils');
var validation = require('./validation');

var fields = {
  // these fields are part of the public interface
  TRANSACTION_ID: {
    name: 'transactionID',
    displayName: 'Transaction ID',
    startsAt: 149,
    length: 32,
    fnRead: bufferUtil.read.fixedLenUtf8String
  },
  ROOT_CONTEXT_ID: {
    name: 'rootContextID',
    displayName: 'Root Context ID',
    startsAt: 186,
    length: 16,
    fnRead: bufferUtil.read.fixedLenHexString
  },
  PREVIOUS_COMPONENT: {
    name: 'previousComponent',
    displayName: 'Previous Component',
    startsAt: 117,
    length: 32,
    fnValidate: validation.validatePreviousComponent,
    fnWrite: bufferUtil.write.variableLenUtf8String
  },
  CONNECTION_ID: {
    name: 'connectionID',
    displayName: 'Connection ID',
    startsAt: 202,
    length: 16,
    fnRead: bufferUtil.read.fixedLenHexString,
    fnValidate: validation.validateConnectionID,
    fnWrite: bufferUtil.write.fixedLenHexString
  },
  CONNECTION_COUNTER: {
    name: 'connectionCounter',
    displayName: 'Connection Counter',
    startsAt: 218,
    length: 4,
    fnRead: bufferUtil.read.uintBigEndian,
    fnValidate: validation.validateConnectionCounter,
    fnWrite: bufferUtil.write.uintBigEndian
  },

  // these fields are for internal use
  VAR_PARTS_COUNTER: {
    startsAt: 222,
    length: 2
  },
  VAR_PARTS_OFFSET: {
    length: 2
  },
  PASSPORT_LENGTH: {
    startsAt: 5,
    length: 2,
    fnWrite: bufferUtil.write.uintBigEndian
  }
};

var readableFields = [
  fields.TRANSACTION_ID,
  fields.ROOT_CONTEXT_ID,
  fields.CONNECTION_ID,
  fields.CONNECTION_COUNTER
];

var writableFields = [
  fields.PREVIOUS_COMPONENT,
  fields.CONNECTION_ID,
  fields.CONNECTION_COUNTER
];

module.exports = {
  fields: fields,
  readableFields: readableFields,
  writableFields: writableFields
};
