// @sap/audit-logging/lib/v1/DataAccessMessage.js
'use strict';

var utils = require('../utils');
var common = require('./common');
var DataAccessBase = require('../base-messages/DataAccessBase');

module.exports = DataAccessMessage;

function DataAccessMessage(content, transport) {
  DataAccessBase.call(this, content, transport);
}

require('util').inherits(DataAccessMessage, DataAccessBase);

DataAccessMessage.prototype.attachment = function (id, name) {
  utils.validate.notEmptyString(id, 'id');
  if (name !== undefined) {
    utils.validate.notEmptyString(name, 'name');
  }

  if (!this._content.attachments) {
    this._content.attachments = [];
  }

  this._content.attachments.push({
    id: id,
    name: name
  });

  return this;
};

DataAccessMessage.prototype.attribute = function (name, successful) {
  utils.validate.notEmptyString(name, 'attribute name');
  if (successful !== undefined) {
    utils.validate.boolean(successful, 'attribute successful');
  }

  utils.validate.attributeName(this._content.attributes, name);

  this._content.attributes.push({
    name: name,
    successful: successful
  });

  return this;
};

DataAccessMessage.prototype.log = common.log;
