// @sap/approuter/lib/utils/url-utils.js
'use strict';

/**
 * We are not using the built-in url module, because when it parse
 * an url, it escapes some characters like quoutes and asteriks, but we need the url
 * to be forwarded as it came to us.
 */
var _ = require('lodash');
var urijs = require('urijs');

module.exports = {
  parse: function (urlString) {
    var parsedUri = urijs.parse(urlString);

    var resultUrl = {
      protocol: parsedUri.protocol ? parsedUri.protocol + ':' : null,
      hostname: parsedUri.hostname || null,
      host: buildHost(parsedUri),
      port: parsedUri.port ? parsedUri.port : null,
      pathname: parsedUri.path,
      slashes: true,
      auth: buildUserInfo(parsedUri),
      search: parsedUri.query ? ('?' + parsedUri.query) : null,
      query: parsedUri.query ? parsedUri.query : null,
      hash: parsedUri.fragment ? ('#' + parsedUri.fragment) : null
    };

    resultUrl.path = resultUrl.pathname + (resultUrl.search || '');

    resultUrl.href = resultUrl.protocol ? (resultUrl.protocol + '//') : '';
    resultUrl.href += resultUrl.auth ? (resultUrl.auth + '@') : '';
    resultUrl.href += resultUrl.host ? resultUrl.host : '';
    resultUrl.href += resultUrl.path;
    resultUrl.href += resultUrl.hash ? resultUrl.hash : '';

    return resultUrl;
  },

  buildAppRouterUrl: function (req) {
    var reqHeaders = req.headers;
    var forwardedProtocol = reqHeaders['x-forwarded-proto'];
    var host = getRedirectHost(req);

    if (forwardedProtocol) {
      return forwardedProtocol + '://' + host;
    }

    var hostHeaderParts = host.split(':');
    if (hostHeaderParts.length === 2) {
      var port = hostHeaderParts[1];
      if (port === '443') {
        return 'https://' + host;
      }
      if (port === '80') {
        return 'http://' + host;
      }
    }
    return req.protocol + '://' + host;
  },

  getAppRouterHost: getAppRouterHost,
  getRedirectHost: getRedirectHost,

  join: function (url, path) {
    return _.trimEnd(url, '/') + '/' + _.trimStart(path, '/');
  }
};

function getRedirectHost(req) {
  var reqHeaders = req.headers;
  return reqHeaders['x-forwarded-host'] || reqHeaders.host;
}

function getAppRouterHost(req) {
  var reqHeaders = req.headers;
  var extReverseProxy = req.routerConfig.externalReverseProxy;
  if (extReverseProxy){
    return reqHeaders['x-custom-host'] ||  reqHeaders.host;
  }
  return reqHeaders['x-forwarded-host'] ||  reqHeaders.host;
}

function buildHost(parsedUri) {
  var host = parsedUri.hostname ? parsedUri.hostname : null;
  if (host && parsedUri.port) {
    host += ':' + parsedUri.port;
  }
  return host;
}

function buildUserInfo(parsedUri) {
  var res = '';
  if (typeof parsedUri.username === 'string') {
    res = parsedUri.username;
  }
  if (typeof parsedUri.password === 'string') {
    res += ':' + parsedUri.password;
  }
  return res || null;
}
