// @sap/approuter/lib/passport/jwt-refresh.js
'use strict';

var VError = require('verror').VError;
var loggerUtils = require('../utils/logger');
var passportUtils = require('./utils');

var tracer = loggerUtils.getTracer(__filename);
var logger = loggerUtils.getLogger('/request/uaa');

module.exports = {
  executeAfter: function (fn, timeout) {
    setTimeout(fn, timeout);
  },

  refreshToken: function (app, session) {
    tracer.path('Entering refreshToken');

    var routerConfig = app.get('mainRouterConfig');
    var jwtRefresh = routerConfig.jwtRefresh;
    var tokenRefreshTimestamp = session.user.token.expiryDate - toMilliseconds(jwtRefresh);

    var msBeforeRefresh = tokenRefreshTimestamp - Date.now();
    if (msBeforeRefresh < 0) {
      return tracer.error('jwtRefresh is greater than the the token validity period!');
    }

    var memoryStore = app.get('memoryStore');
    memoryStore.getSessionTimeout(session.id, function (err, sessionTimeout) {
      if (err || Date.now() + toMilliseconds(sessionTimeout) < tokenRefreshTimestamp) {
        return err && tracer.debug(err);
      }

      memoryStore.update(session.id, function (newSession) {
        if (!newSession || newSession.jwtRefreshStarted) {
          return;
        }
        tracer.info('Updating jwtRefreshStarted to true for session %s.', newSession.id);
        newSession.jwtRefreshStarted = true;
        tracer.info('Refreshing jwt for session %s in %dms.', newSession.id, msBeforeRefresh);
        module.exports.executeAfter(function () {
          doRefresh(app, newSession);
        }, msBeforeRefresh);
      }, false);
    });
  }
};

function toMilliseconds(minutes) {
  return minutes * 60 * 1000;
}

function doRefresh(app, session) {
  tracer.path('Entering doRefresh');
  var oauthOptions = session.user.token.oauthOptions;
  var requestOptions = {
    url: oauthOptions.tokenURL,
    auth: {
      user: oauthOptions.clientid,
      pass: oauthOptions.clientsecret
    },
    form: {
      'grant_type': 'refresh_token',
      'refresh_token': session.user.token.refreshToken
    }
  };
  requestOptions.timeout = session.user.token.expiryDate - Date.now();
  var checkTimeout = function () {
    if (session.user.token.expiryDate - Date.now() < 0) {
      tracer.info('JWT for session %s could not be refreshed!', session.id);
      return new VError('JWT could not be refreshed,it is expired!');
    }
  };
  passportUtils.callUaa(requestOptions, checkTimeout, function (err, uaaResponse) {
    if (err) {
      tracer.warning(err, 'Session %s', session.id);
      return logger.error('Unable to refresh the JWT for session %s , number of attempts done: %d, error: %s', session.id, passportUtils.MAX_ATTEMPTS_NUMBER, err.message);
    }
    if (!uaaResponse.access_token || !uaaResponse.expires_in || !uaaResponse.refresh_token) {
      return logger.error('Bad response from UAA when refreshing JWT - not all JWT fields are present');
    }
    var updatedSession;
    app.get('memoryStore').update(session.id, function updateSession(newSession) {
      if (!newSession) {
        return;
      }
      var options = {
        accessToken: uaaResponse.access_token,
        expiresIn: uaaResponse.expires_in,
        refreshToken: uaaResponse.refresh_token,
        scope: uaaResponse.scope,
        oauthOptions: newSession.user.token.oauthOptions
      };
      newSession.user = passportUtils.getUserProperties(options);
      newSession.jwtRefreshStarted = false;
      updatedSession = newSession;
      tracer.info('Token refreshed!');
    }, false);
    updatedSession && module.exports.refreshToken(app, updatedSession); // schedule next refresh if needed
  });
}
